addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java

```diff
@@ -143,50 +159,112 @@ private void changeState(final InternalLedState newState) throws IOException {
         }
     }
 
-    private void fadeToState(final InternalLedState newState) throws IOException {
-        if (!newState.equals(realTargetState)) {
-            keepFading = false;
-            realTargetState = newState;
+    final static class LEDFaderRunner implements Runnable {
+        protected Logger logger = LoggerFactory.getLogger(LEDFaderRunner.class);
+        InternalLedState fadeState;
+        int fadeStepper = 1;
+
+        final InternalLedState toState;
+        final InternalLedState fromState;
+        final Future<?>[] faderArray;
+        final int steps;
+        final Function<InternalLedState, Boolean> ledSender;
+
+        public LEDFaderRunner(Future<?>[] faderFuture, InternalLedState fromState, InternalLedState toState, int steps, Function<InternalLedState,  Boolean> ledSender) {
+            this.faderArray = faderFuture;
+            fadeState=fromState;
+            this.fromState = fromState;
+            this.toState = toState;
+            this.steps = steps;
+            this.ledSender = ledSender;
+        }
 
-            executorService.schedule(() -> {
-                if (currentTargetState.equals(newState)) {
-                    return;
+        public void run() {
+            Future<?> thisFuture;
+            while (null == (thisFuture = faderArray[0])) Thread.yield();
+            if (fadeStepper <= steps) {
+                if (!Thread.currentThread().isInterrupted()) {
+                    fadeState = fromState.fade(toState, (double) fadeStepper / steps);
+                    logger.debug("fadeState: {}", fadeState);
+                    if (!ledSender.apply(fadeState)) {
+                        logger.debug("fadeState: sending failed at step {}", fadeStepper);
+                        thisFuture.cancel(true);
+                    }
                 }
+            } else {
+                thisFuture.cancel(true);
```

Comment:
If I understand well this happens only in case of a miscalculation or bug (or maybe a rounding error)?